### PR TITLE
[IVY-3035] Fix Exchange Rates not Saving to Database

### LINKED ISSUE #3035: Sync of exchange rates not saving to db
### Please confirm the following

- [X] I checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate problems

### Describe the bug

The Syncing of exchange rates are not working if an account of different currency is used

### To Reproduce

1. Create an Account with a different Currency which is not the same as base currency 
2. add expense and income transactions 
3. balance is broken

### Expected behavior

_No response_

### Screenshots

_No response_

### Smartphone

Pixel 7a (Emulator)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Vishwa-Raghavendra**
## Pull Request (PR) Checklist
Please check if your pull request fulfills the following requirements:
- [x] The PR is submitted to the `main` branch.
- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read the [Architecture Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Architecture.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] 🎬 I've attached a **screen recoding** of the changes. 

> Tip: drag & drop the video to the PR description.

## What's changed?

Describe with a few bullets **what's new:**
- a Fixes Exchange Rates Not Saving to Database
- b
- c

> 💡 Tip: Please, attach screenshots and screen recordings. It helps a lot!

## Risk Factors

**What may go wrong if we merge your PR?**

- a
- b

**In what cases your code won't work?**

- a
- b

## Does this PR closes any GitHub Issues?

Check **[Ivy Wallet Issues](https://github.com/Ivy-Apps/ivy-wallet/issues)**.

- Closes #3035 #3002 

> Replace `ISSUE_NUMBER` with the number of the GitHub issue that this PR is fixing. If you've done that correctly, you'll see the issue title linked when previewing your PR description.

> **!Note:** Do **not** link the PR number. Link the number/id of the GitHub Issue from [issues](https://github.com/Ivy-Apps/ivy-wallet/issues).

## Troubleshooting CI failures ❌

GitHub Actions failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

File: shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt [hunk 1]
```diff
@@ -1,28 +1,41 @@
...
 import io.ktor.client.HttpClient
 import io.ktor.client.call.body
 import io.ktor.client.request.get
+import java.util.Locale
 import javax.inject.Inject
 
 class RemoteExchangeRatesDataSourceImpl @Inject constructor(
     private val ktorClient: dagger.Lazy<HttpClient>,
 ) : RemoteExchangeRatesDataSource {
+    private val currencyCodeLowerMask: String = "###"
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt:16]: Why do we need this? Let's hardcode the URLs to eur because EUR will always be available as a base and the code is simpler

File: shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt [hunk 2]
```diff
@@ -1,11 +1,62 @@
...
+import kotlinx.serialization.json.JsonTransformingSerializer
+import kotlinx.serialization.json.contentOrNull
+import kotlinx.serialization.json.encodeToJsonElement
+import kotlinx.serialization.json.jsonPrimitive
 
-@Serializable
+@Serializable(with = ExchangeRatesResponseDeserializer::class)
 @Keep
 data class ExchangeRatesResponse(
+    @SerialName("date")
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:23]: I think this is the real fix but why do we need a custom serializer? It's again complex 😂 Before when I reworked it to Ktor it worked w/o a custom serializer. Just `@SerialName`

File: shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt (hunk 1 above)
Comment (Vishwa-Raghavendra) [reply shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt:16]: We need it because the URL Changes based on the currency Code 
eg ->  This is for EUR
https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies/eur.min.json 

If I want USD then 
https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies/usd.min.json

the API path itself changes rather than providing a query parameter hence we need this

File: shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt (hunk 2 above)
Comment (Vishwa-Raghavendra) [reply shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:23]: we need a custom serializer because the JSON is dynamic and there is no easy way to handle it kotlinx.serialization 
if it was jackson I coud have used @JsonAnySetter instead and fixed it up 

Dynamic Json e.g --> if Currency is USD -->
{
    "date": "2024-03-24",
    "usd": {
        "$myro": 4.62721175,
        "$wen": 3744.01152636,
        .....
        }
}

If currency is EUR  -->
{
    "date": "2024-03-24",
    "eur": {
        "$myro": 4.62721175,
        "$wen": 3744.01152636,
        .....
        }
}

We are trying to get it in 

{
date: "..."
rates: "..."
}
so rates as a json field will never be available

Comment (ILIYANGERMANOV) [reply shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:23]: Can't we just rename "rates" to "eur" in the SerialName and call it a day? We don't need that extra complexity. It creates new codepaths that can fail with no added benefit

Comment (ILIYANGERMANOV) [reply shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:23]: We don't neeb dynamic base currency. Fetch everything in EUR and then recompute to user's base currency. This computation is almost instant

Comment (Vishwa-Raghavendra) [reply shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:23]: If we were only fetching eur that would be fine but lets say we fetch USD 
instead of "eur" you will get "usd" in json response

File: shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt (hunk 1 above)
Comment (ILIYANGERMANOV) [reply shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt:16]: No we don't - just fetch EUR which I know it works in our free limited open source API which is btw some json files in a GitHub repo. Always fetch EUR -> recompute to user's base

File: shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt (hunk 2 above)
Comment (ILIYANGERMANOV) [reply shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:23]: Yes, I agree but let's fetch only EUR for the sake of simplicity and API stability. I bet our API doesn't support RSD as a base

File: shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt (hunk 1 above)
Comment (Vishwa-Raghavendra) [reply shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt:16]: Got it

**Review by ILIYANGERMANOV — APPROVED** [review]
Awesome!! Thank you for finding and fixing this 🏆 I love this simple fix haha. I wonder why our Integration tests haven't caught that bug 😬 

As an improvement, in the future we should add integration tests that fetches the rates and check if the DB is not empty
